packages/opentelemetry-plugin-xml-http-request/src/xhr.ts

```diff
@@ -0,0 +1,371 @@
+      }
+      const currentSpan = spanId ? plugin._spans[spanId] : undefined;
+      if (currentSpan) {
+        plugin._addSpanNetworkEvents(currentSpan);
+        currentSpan.addEvent(eventName);
+
+        const url = currentSpan.attributes[AttributeNames.HTTP_URL];
+        if (typeof url === 'string') {
+          const parsedUrl = parseUrl(url);
+
+          currentSpan.setAttribute(AttributeNames.HTTP_STATUS_CODE, xhr.status);
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_STATUS_TEXT,
+            xhr.statusText
+          );
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_HOSTNAME,
+            parsedUrl.hostname
+          );
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_PATH,
+            parsedUrl.pathname
+          );
+
+          // @TODO do we want to collect this or it will be collected earlier once only or
+          //    maybe when parent span is not available ?
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_USER_AGENT,
+            navigator.userAgent
+          );
+        }
+
+        currentSpan.end();
+      }
+    }
+
+    function onError(this: XMLHttpRequestWrapped) {
+      endSpan(this, EventNames.EVENT_ERROR);
+    }
+
+    function onTimeout(this: XMLHttpRequestWrapped) {
+      endSpan(this, EventNames.EVENT_TIMEOUT);
+    }
+
+    function onLoad(this: XMLHttpRequestWrapped) {
+      if (this.status < 299) {
+        endSpan(this, EventNames.EVENT_LOAD);
+      } else {
+        endSpan(this, EventNames.EVENT_ERROR);
+      }
+    }
+
+    function unregister(xhr: XMLHttpRequestWrapped, spanId: string) {
+      xhr.removeEventListener('load', onLoad);
+      xhr.removeEventListener('error', onError);
+      xhr.removeEventListener('timeout', onTimeout);
+      delete plugin._callbackToRemoveEvents[spanId];
+      delete xhr.__OT_SPAN_ID;
+    }
+
+    return (original: SendFunction): SendFunction => {
+      return function patchSend(
+        this: XMLHttpRequest,
+        body?: SendBody
+      ): SendFunction {
+        const spanId = (this as XMLHttpRequestWrapped).__OT_SPAN_ID;
+        const currentSpan: types.Span | undefined = spanId
+          ? plugin._spans[spanId]
+          : undefined;
+
+        if (currentSpan && spanId) {
+          currentSpan.addEvent(EventNames.METHOD_SEND);
+
+          this.addEventListener('load', onLoad);
+          this.addEventListener('error', onError);
+          this.addEventListener('timeout', onTimeout);
+
+          plugin._callbackToRemoveEvents[spanId] = () => {
+            unregister(this as XMLHttpRequestWrapped, spanId);
+          };
+          plugin._addHeaders(this, currentSpan);
+        }
+        return original.call(this, body);
+      };
+    };
+  }
+
+  /**
+   * implements patch function
+   */
+  protected patch() {
+    this._logger.debug('applying patch to', this.moduleName, this.version);
+
+    if (isWrapped(XMLHttpRequest.prototype.open)) {
```

Comment:
with web there is always some probability that something can be destroyed and recreated again. This way it will always work correctly

Reply:
Since this is patching a global, would it not patch if multiple components on the same page are instrumented with separate tracers/configs?

Reply:
copy paste here my reply from below
We are using shimmer which doesn't support multiple wrap of the same function so if you even register multiple tracers with the same plugins, and try to unpatch just one, it will unpatch the global successfully anyway. So if the decision to use shimmer was done my assumption is that we are able to instrument using the plugins once only. Otherwise the code will be broken anyway.